### PR TITLE
fix: Fix cloud storage state backend prefix duplication bug

### LINKED ISSUE #7938: bug: Unexpected behaviour from S3 State Backend
### Meltano Version

2.19.0

### Python Version

3.8

### Bug scope

API

### Operating System

Mac OS 13.4.1

### Description

Attempting to run a pipeline with S3 State Backend set.

Two strange issues occurred.

1. When I set `MELTANO_STATE_BACKEND_URI='s3://my-data-lake/state/` the files are saved to `s3://my-data-lake/state/state/<env>:<pipeline name>/`  so whatever subdirectory I use in my key is captured and reused as part of the key. It looks like the `/state/` part of my URI is correctly being parsed out as a prefix but is not being removed from the base URI before the request is sent to S3.
2. If I delete the state files or change the URI and attempt to rerun a pipeline the boto3 delete object/retry code seems to b0rk and stops any state being saved. Expected behaviour is a graceful exit or a run with the state saved on S3.
Pipeline run _without_  `--full-refresh` it should just assume it is the first time as there are no state files available at the S3 Backend

For the second case the process stops like this:
```
2023-07-20T16:30:12.135783Z [info     ] Reading state from AWS S3
2023-07-20T16:30:12.149455Z [info     ] Found credentials in environment variables.
2023-07-20T16:30:12.748605Z [error    ] Cannot start plugin tap-mysql: argument of type 'int' is not iterable
2023-07-20T16:30:12.752616Z [error    ] Block run completed.           block_type=ExtractLoadBlocks err=RunnerError("Cannot start plugin tap-mysql: argument of type 'int' is not iterable") exit_codes={} set_number=0 success=False
```

It seems that whatever result is passed back to the tap is not usable and breaks the pipeline.

### Code

```python
1. State Backend S3

.env

MELTANO_STATE_BACKEND_URI='s3://my-data-lake/state/'

Log:

2023-07-21T18:01:30.162987Z [debug    ] Response body:
b'<?xml version="1.0" encoding="UTF-8"?>\n\n<CompleteMultipartUploadResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Location>https://my-data-lake.s3.eu-west-1.amazonaws.com/state%2Fstate%2Fmy-test%3Atap-mysql-to-target-s3%2Fstate.json</Location><Bucket>my-data-lake</Bucket><Key>state/state/my-test:tap-mysql-to-target-s3/state.json</Key><ETag>"e96596dc571d5a2a2193055ec786726f-1"</ETag></CompleteMultipartUploadResult>'
2023-07-21T18:01:30.163329Z [debug    ] Event needs-retry.s3.CompleteMultipartUpload: calling handler <function check_for_200_error at 0x10ab908b0>
2023-07-21T18:01:30.163476Z [debug    ] Event needs-retry.s3.CompleteMultipartUpload: calling handler <botocore.retryhandler.RetryHandler object at 0x10bce0580>
2023-07-21T18:01:30.163608Z [debug    ] No retry needed.
2023-07-21T18:01:30.163712Z [debug    ] Event needs-retry.s3.CompleteMultipartUpload: calling handler <bound method S3RegionRedirectorv2.redirect_from_error of <botocore.utils.S3RegionRedirectorv2 object at 0x10bce05e0>>
2023-07-21T18:01:30.163874Z [debug    ] smart_open.s3.MultipartWriter('my-data-lake', 'state/state/my-test:tap-mysql-to-target-s3/state.json'): completed multipart upload


2. Attempted (re)run with changed MELTANO_STATE_BACKEND_URI or missing state files and without --full-refresh


2023-07-20T16:31:46.835228Z [debug    ] Event request-created.s3.DeleteObjects: calling handler <function add_retry_headers at 0x110e66a60>
2023-07-20T16:31:46.839799Z [debug    ] Sending http request: <AWSPreparedRequest stream_output=False, method=POST, url=https://my-data-lake.s3.eu-west-1.amazonaws.com/?delete, headers={'User-Agent': b'Boto3/1.28.7 md/Botocore#1.31.7 ua/2.0 os/macos#22.5.0 md/arch#x86_64 lang/python#3.8.16 md/pyimpl#CPython cfg/retry-mode#legacy Botocore/1.31.7', 'Content-MD5': b'X+**==', 'X-Amz-Date': b'20230720T163146Z', 'X-Amz-Security-Token': b'***Q=', 'X-Amz-Content-SHA256': ****', 'Authorization': b'AWS4-HMAC-SHA256 Credential=**/**/eu-west-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date;x-amz-security-token, Signature=***', 'amz-sdk-invocation-id': b'***-101c-462e-a487-**', 'amz-sdk-request': b'attempt=1', 'Content-Length': '132'}>
2023-07-20T16:31:46.841607Z [debug    ] Certificate path: /****/venv/lib/python3.8/site-packages/certifi/cacert.pem
2023-07-20T16:31:47.039194Z [debug    ] Response headers: {'x-amz-id-2': 'uG6q8J/Em8uuGjQu4aiK4+****==', 'x-amz-request-id': '*****', 'Date': 'Thu, 20 Jul 2023 16:31:48 GMT', 'Content-Type': 'application/xml', 'Transfer-Encoding': 'chunked', 'Server': 'AmazonS3', 'Connection': 'close'}
2023-07-20T16:31:47.039463Z [debug    ] Response body:
b'<?xml version="1.0" encoding="UTF-8"?>\n<DeleteResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Deleted><Key>/my-test:tap-mysql-to-target-s3/lock</Key><DeleteMarker>true</DeleteMarker><DeleteMarkerVersionId>****</DeleteMarkerVersionId></Deleted></DeleteResult>'
2023-07-20T16:31:47.039996Z [debug    ] Event needs-retry.s3.DeleteObjects: calling handler <botocore.retryhandler.RetryHandler object at 0x111f4ee80>
2023-07-20T16:31:47.040135Z [debug    ] No retry needed.
2023-07-20T16:31:47.040249Z [debug    ] Event needs-retry.s3.DeleteObjects: calling handler <bound method S3RegionRedirectorv2.redirect_from_error of <botocore.utils.S3RegionRedirectorv2 object at 0x111f4eee0>>
2023-07-20T16:31:47.040900Z [debug    ] Deleted configuration at ****/.meltano/run/tap-mysql/tap.6aaac143-93b4-4bb6-9e6c-d7bd8e086b3b.config.json
2023-07-20T16:31:47.044297Z [error    ] Cannot start plugin tap-mysql: argument of type 'int' is not iterable
2023-07-20T16:31:47.048569Z [error    ] Block run completed.           block_type=ExtractLoadBlocks err=RunnerError("Cannot start plugin tap-mysql: argument of type 'int' is not iterable") exit_codes={} set_number=0 success=False
2023-07-20T16:31:50.147014Z [debug    ] Need help fixing this problem? Visit http://melta.no/ for troubleshooting steps, or to
join our friendly Slack community.

Run invocation could not be completed as block failed: Cannot start plugin tap-mysql: argument of type 'int' is not iterable 
Traceback (most recent call last):
  File "/****/venv/lib/python3.8/site-packages/meltano/core/state_store/filesystem.py", line 202, in is_locked
    with self.get_reader(lock_path) as reader:
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/****/venv/lib/python3.8/site-packages/meltano/core/state_store/filesystem.py", line 90, in get_reader
    with open(
  File "/****/venv/lib/python3.8/site-packages/smart_open/smart_open_lib.py", line 177, in open
    fobj = _shortcut_open(
  File "/****/venv/lib/python3.8/site-packages/smart_open/smart_open_lib.py", line 363, in _shortcut_open
    return _builtin_open(local_path, mode, buffering=buffering, **open_kwargs)
FileNotFoundError: [Errno 2] No such file or directory: '/my-test:tap-mysql-to-target-s3/lock'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/****/venv/lib/python3.8/site-packages/meltano/core/block/singer.py", line 336, in start
    self.process_handle = await self.invoker.invoke_async(
  File "/****/venv/lib/python3.8/site-packages/meltano/core/plugin_invoker.py", line 443, in invoke_async
    async with self._invoke(*args, **kwargs) as (
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/contextlib.py", line 171, in __aenter__
    return await self.gen.__anext__()
  File "/****/venv/lib/python3.8/site-packages/meltano/core/plugin_invoker.py", line 419, in _invoke
    async with self.plugin.trigger_hooks("invoke", self, args):
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/contextlib.py", line 171, in __aenter__
    return await self.gen.__anext__()
  File "/****/venv/lib/python3.8/site-packages/meltano/core/behavior/hookable.py", line 92, in trigger_hooks
    await self.__class__.trigger(self, f"before_{hook_name}", *args, **kwargs)
  File "/****/venv/lib/python3.8/site-packages/meltano/core/behavior/hookable.py", line 120, in trigger
    raise err
  File "/****/venv/lib/python3.8/site-packages/meltano/core/behavior/hookable.py", line 112, in trigger
    await hook_func(target, *args, **kwargs)
  File "/****/venv/lib/python3.8/site-packages/meltano/core/plugin/singer/tap.py", line 254, in look_up_state_hook
    await self.look_up_state(plugin_invoker)
  File "/****/venv/lib/python3.8/site-packages/meltano/core/plugin/singer/tap.py", line 311, in look_up_state
    state = StateService(
  File "/****/venv/lib/python3.8/site-packages/meltano/core/state_service.py", line 158, in get_state
    state = self.state_store_manager.get(state_id=state_id)
  File "/****/venv/lib/python3.8/site-packages/meltano/core/state_store/filesystem.py", line 274, in get
    with self.acquire_lock(state_id):
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/****/venv/lib/python3.8/site-packages/meltano/core/state_store/filesystem.py", line 241, in acquire_lock
    while self.is_locked(state_id):
  File "/****/venv/lib/python3.8/site-packages/meltano/core/state_store/filesystem.py", line 211, in is_locked
    if self.is_file_not_found_error(e):
  File "/****/venv/lib/python3.8/site-packages/meltano/core/state_store/s3.py", line 92, in is_file_not_found_error
    and "NoSuchKey" in err.args[0]
TypeError: argument of type 'int' is not iterable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/run.py", line 172, in _run_blocks
    await blk.run()
  File "/****/venv/lib/python3.8/site-packages/meltano/core/block/extract_load.py", line 444, in run
    await self.run_with_job()
  File "/****/venv/lib/python3.8/site-packages/meltano/core/block/extract_load.py", line 474, in run_with_job
    await self.execute()
  File "/****/venv/lib/python3.8/site-packages/meltano/core/block/extract_load.py", line 433, in execute
    async with self._start_blocks():
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/contextlib.py", line 171, in __aenter__
    return await self.gen.__anext__()
  File "/****/venv/lib/python3.8/site-packages/meltano/core/block/extract_load.py", line 563, in _start_blocks
    await block.start()
  File "/****/venv/lib/python3.8/site-packages/meltano/core/block/singer.py", line 343, in start
    raise RunnerError(f"Cannot start plugin {self.string_id}: {err}") from err
meltano.core.runner.RunnerError: Cannot start plugin tap-mysql: argument of type 'int' is not iterable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/__init__.py", line 105, in _run_cli
    cli(obj={"project": None})
  File "/****/venv/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/cli.py", line 43, in main
    return super().main(*args, windows_expand_args=False, **kwargs)
  File "/****/venv/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/utils.py", line 612, in invoke
    super().invoke(ctx)
  File "/****/venv/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/utils.py", line 661, in invoke
    super().invoke(ctx)
  File "/****/venv/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/****/venv/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/params.py", line 27, in decorate
    return func(*args, **kwargs)
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/params.py", line 76, in decorate
    func(project, *args, **kwargs)
  File "/****/venv/lib/python3.8/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/****/venv/lib/python3.8/site-packages/meltano/core/utils/__init__.py", line 79, in wrapper
    return asyncio.run(func(*args, **kwargs))
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/run.py", line 141, in run
    raise err
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/run.py", line 138, in run
    await _run_blocks(tracker, parsed_blocks, dry_run=dry_run)
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/run.py", line 184, in _run_blocks
    raise CliError(
meltano.cli.utils.CliError: Run invocation could not be completed as block failed: Cannot start plugin tap-mysql: argument of type 'int' is not iterable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/****/venv/lib/python3.8/site-packages/meltano/cli/__init__.py", line 115, in _run_cli
    raise CliError(f"{troubleshooting_message}\n{err}") from err
meltano.cli.utils.CliError: Need help fixing this problem? Visit http://melta.no/ for troubleshooting steps, or to
join our friendly Slack community.

Run invocation could not be completed as block failed: Cannot start plugin tap-mysql: argument of type 'int' is not iterable
```

### PR DESCRIPTION
**Pull request by cjohnhanson**
See [this comment](https://github.com/meltano/meltano/issues/7938#issuecomment-1650632775) for explanation of the root cause and discussion of motivation for the `UpgradeService` changes here. 

### Testing strategy
Replicate bug with each cloud state backend using latest release; do editable install of this branch then manually run `UpgradeService.migrate_state`; confirm that state files were copied to deduplicated paths; run pipeline again and confirm that state is found at new path.

Above has been tested with:
- [X] S3
- [x] GCS
- [X] Azure (using [Azurite](https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azurite?tabs=visual-studio)

### CODE REVIEW

**Review by WillDaSilva — APPROVED** [review]
Sorry for the delayed review @cjohnhanson. This fell off my radar.

**Comment by edgarrmondragon** [comment]
I don't think `UpgradeService.migrate_state` is called anywhere, e.g. in the `meltano upgrade` CLI. Should we have that before merging?

**Comment by cjohnhanson** [comment]
@edgarrmondragon 
> I don't think UpgradeService.migrate_state is called anywhere, e.g. in the meltano upgrade CLI. Should we have that before merging?

That was a pretty big oversight, thanks for catching that. 

Any tips/ideas on testing the end-to-end upgrade? My manual testing approach tested the `upgrade_state` method on its own and I wouldn't anticipate there being any issues but would feel much more comfortable if an end-to-end upgrade with state changes could be tested as well.

**Comment by edgarrmondragon** [comment]
> @edgarrmondragon
> 
> > I don't think UpgradeService.migrate_state is called anywhere, e.g. in the meltano upgrade CLI. Should we have that before merging?
> 
> That was a pretty big oversight, thanks for catching that.
> 
> Any tips/ideas on testing the end-to-end upgrade? My manual testing approach tested the `upgrade_state` method on its own and I wouldn't anticipate there being any issues but would feel much more comfortable if an end-to-end upgrade with state changes could be tested as well.

@cjohnhanson There's [`tests/meltano/cli/test_upgrade.py`](https://github.com/meltano/meltano/blob/40b4a8cf1516dbbe4cfac0acc579bffb5e90a45c/tests/meltano/cli/test_upgrade.py), so maybe adding a test there and use [moto](https://docs.getmoto.org/en/latest/docs/services/s3.html#s3) to mock S3, adding a file as it'd be before upgrading, then running the upgrade, then checking the file was correctly migrated?

File: src/meltano/core/upgrade_service.py [hunk 1]
```diff
@@ -174,7 +176,36 @@ def migrate_database(self):
...
+                    )
+                    new_path = filepath.replace(duplicated_substr, manager.prefix)
+                    new_path = new_path.replace(
+                        manager.delimiter * 2,
+                        manager.delimiter,
+                    )
+                    manager.copy_file(filepath, new_path)
+                    click.secho(f"Copied state from {filepath} to {new_path}")
+
+    def upgrade(self, skip_package: bool = False, **kwargs):  # noqa: WPS213
```
Comment (edgarrmondragon) [inline src/meltano/core/upgrade_service.py:208]: Confusingly, I don't think `UpgradeService.upgrade` this is called anywhere. You probably want to call `UpgradeService.migrate_state` in https://github.com/meltano/meltano/blob/b88f3fc311bd94cf4fb5e0edb652ba386fae9fae/src/meltano/cli/upgrade.py#L80-L85

File: tests/meltano/cli/test_upgrade.py [hunk 2]
```diff
@@ -197,3 +199,24 @@ def test_upgrade_files_glob_path(self, project, cli_runner):
...
+            )
+        monkeypatch.setenv(
+            "MELTANO_STATE_BACKEND_URI",
+            "s3://test-state-bucket/some/trailing/delim/path/",
+        )
+        result = cli_runner.invoke(cli, ["upgrade"])
+        assert_cli_runner(result)
+        keys = [s3_object.key for s3_object in bucket.objects.all()]
+        for state_id in state_ids:
+            key = f"some/trailing/delim/path/{state_id}/state.json"
```
Comment (edgarrmondragon) [inline tests/meltano/cli/test_upgrade.py:221]:
```suggestion
            key = f"/some/trailing/delim/path/{state_id}/state.json"
```

File: src/meltano/core/upgrade_service.py (hunk 1 above)
Comment (cjohnhanson) [reply src/meltano/core/upgrade_service.py:208]: Ah, weird. I'll add it there as well. I'll also go ahead and add a comment in `UpgradeService.upgrade` to indicate that it's not called and can probably be deprecated in the future.

File: tests/meltano/cli/test_upgrade.py [hunk 3]
```diff
@@ -197,3 +199,24 @@ def test_upgrade_files_glob_path(self, project, cli_runner):
...
+        bucket = conn.create_bucket(Bucket="test-state-bucket")
+        for state_id in state_ids:
+            bucket.put_object(
+                Key=f"some/trailing/delim/path/some/trailing/delim/path/{state_id}/state.json",  # noqa: E501
+            )
+        monkeypatch.setenv(
+            "MELTANO_STATE_BACKEND_URI",
+            "s3://test-state-bucket/some/trailing/delim/path/",
+        )
+        result = cli_runner.invoke(cli, ["upgrade"])
```
Comment (edgarrmondragon) [inline tests/meltano/cli/test_upgrade.py:217]:
```suggestion
        result = cli_runner.invoke(cli, ["upgrade", "--skip-package"])
```

File: src/meltano/cli/upgrade.py [hunk 4]
```diff
@@ -85,6 +85,7 @@ def all(ctx, pip_url, force, skip_package):
 
         click.echo()
 
+        upgrade_service.upgrade_state()
```
Comment (edgarrmondragon) [inline src/meltano/cli/upgrade.py:88]:
```suggestion
        upgrade_service.migrate_state()
```

**Review by edgarrmondragon — APPROVED** [review]

**Pull request merged**